packages/story-editor/src/app/story/useStoryReducer/reducers/duplicateElementsById.js

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import { duplicateElement } from '../../../../elements';
+
+/**
+ * Duplicate all elements specified by `elementIds` on the current page.
+ * Set selected elements to be the newly created elements.
+ *
+ * If given `elementIds` are not a list, do nothing.
+ *
+ * @param {Object} state Current state
+ * @param {Object} payload Action payload
+ * @param {Array.<string>} payload.elementIds Array of ids of elements to duplicate.
+ * @return {Object} New state
+ */
+function duplicateElementsById(state, { elementIds }) {
+  if (!Array.isArray(elementIds)) {
+    return state;
+  }
+
+  const pageIndex = state.pages.findIndex(({ id }) => id === state.current);
+  const oldPage = state.pages[pageIndex];
+
+  const newPage = { ...oldPage };
+  const newSelection = [];
+
+  elementIds.forEach((elementId) => {
+    const elementIndex = newPage.elements.findIndex(
+      ({ id }) => id === elementId
+    );
+
+    if (elementIndex < 0) {
+      return;
+    }
+
+    const elementToDuplicate = newPage.elements[elementIndex];
+
+    if (elementToDuplicate.isBackground) {
+      return;
+    }
+
+    const { element, elementAnimations } = duplicateElement({
+      element: elementToDuplicate,
+      animations: oldPage.animations,
+      existingElements: oldPage.elements,
+    });
+
+    newSelection.push(element.id);
+
+    newPage.animations = [...(newPage.animations || []), ...elementAnimations];
+    newPage.elements = [
+      ...newPage.elements.slice(0, elementIndex),
+      elementToDuplicate,
+      element,
+      ...newPage.elements.slice(elementIndex + 1),
+    ];
+
+    return;
+  });
+
```

Comment:
Could we here just `return state` if `!newSelection.length`? Would avoid creating a new history entry if there are no  elements duplicated for some reason.

Reply:
Ah great catch!